// vue-styleguidist/lib/client/rsg-components/Usage/Usage.js
import "core-js/modules/es.object.keys.js";
import React from 'react';
import PropTypes from 'prop-types';
import Props from 'rsg-components/Props';
import Methods from 'rsg-components/Methods';
import Events from 'rsg-components/Events';
import Expose from 'rsg-components/Expose';
import SlotsTable from 'rsg-components/SlotsTable';
export default function Usage(_ref) {
  var _ref$props = _ref.props,
    props = _ref$props.props,
    methods = _ref$props.methods,
    events = _ref$props.events,
    slots = _ref$props.slots,
    expose = _ref$props.expose;
  return /*#__PURE__*/React.createElement("div", null, props ? /*#__PURE__*/React.createElement(Props, {
    props: props
  }) : undefined, methods && methods.length > 0 ? /*#__PURE__*/React.createElement(Methods, {
    methods: methods
  }) : undefined, events && Object.keys(events).length > 0 ? /*#__PURE__*/React.createElement(Events, {
    props: events
  }) : undefined, slots && Object.keys(slots).length > 0 ? /*#__PURE__*/React.createElement(SlotsTable, {
    props: slots
  }) : undefined, expose ? /*#__PURE__*/React.createElement(Expose, {
    expose: expose
  }) : undefined);
}
Usage.propTypes = {
  props: PropTypes.shape({
    props: PropTypes.array,
    methods: PropTypes.array,
    expose: PropTypes.array,
    events: PropTypes.object,
    slots: PropTypes.object
  }).isRequired
};
